### DISCUSSION TITLE
How to perform contouring using masks to isolate isosurface extraction to regions?

### DISCUSSION THREAD

**Discussion by MichaelHillier** [pyvista/pyvista#2107] [Q&A]
Is it possible to perform contouring/marching cubes using boolean masks using pyvista? I am trying to achieve what can be done within skimage.measure.marching_cubes() https://scikit-image.org/docs/stable/api/skimage.measure.html#skimage.measure.marching_cubes . There you can specify a boolean mask where the iso-surfaces will be extracted from within the volume.

I know the contour method within the various volumetric grids (unstructured/structured) does not support mask arrays as input. But I perhaps ghosting the cells could be a way to achieve the same behavior. Anyone know if this is possible? Or another way to achieve iso-surface extraction from particular regions (as specified by a boolean mask array) of a volume?

Thanks

**Comment by banesullivan** [comment]
Using a boolean mask to this end is an interesting idea! I haven't tried this before but I think this is certainly doable - if not, it should be easy enough to use skimage's algorithm with a PyVista mesh. 

Could you provide some sample data or expand on an existing example to show how skimage uses the boolean mask? This might be a good start: https://scikit-image.org/docs/stable/auto_examples/edges/plot_marching_cubes.html#sphx-glr-auto-examples-edges-plot-marching-cubes-py

----

FWIW, you can easily use `skimage.measure.marching_cubes` with PyVista meshes - the beauty of PyVista is its simple numpy interface, making it interoperable with much of the Python data science ecosystem. Here is that example but with PyVista:

```py
import numpy as np

from skimage import measure
from skimage.draw import ellipsoid


# Generate a level set about zero of two identical ellipsoids in 3D
ellip_base = ellipsoid(6, 10, 16, levelset=True)
ellip_double = np.concatenate((ellip_base[:-1, ...],
                               ellip_base[2:, ...]), axis=0)

# Use marching cubes to obtain the surface mesh of these ellipsoids
verts, faces, normals, values = measure.marching_cubes(ellip_double, 0)

vfaces = np.column_stack((np.ones(len(faces),) * 3, faces)).astype(int)

mesh = pv.PolyData(verts, vfaces)
mesh['Normals'] = normals
mesh['values'] = values

mesh.plot(scalars='values')
```
![Screen Shot 2022-02-02 at 9 41 14 PM](https://user-images.githubusercontent.com/22067021/152282318-38cab373-b136-41d9-b397-977b5b832c36.png)

**Reply by MichaelHillier** [reply]
Yes, I am aware we could use both libraries to achieve the goal. Just wondering if it can be done in pyvista. I like have fewer dependencies.

Expanding existing example from skimage to show how to use the boolean masks:
```
import numpy as np
import pyvista as pv
from skimage import measure
from skimage.draw import ellipsoid

# Generate a level set about zero of two identical ellipsoids in 3D
ellip_base = ellipsoid(6, 10, 16, levelset=True)

# Generate boolean arrays for where to extract iso-surface associated to a specific iso-value/level
bool_array1 = np.ones_like(ellip_base, dtype=bool)
bool_array2 = np.ones_like(ellip_base, dtype=bool)
bool_array1[:, :, -16:] = False
bool_array2[:, :, :16] = False

# Generate isosurface 1 from domain defined by bool_array1 using isovalue = 0
verts1, faces1, normals1, values1 = measure.marching_cubes(ellip_base, 0, mask=bool_array1)
vfaces1 = np.column_stack((np.ones(len(faces1),) * 3, faces1)).astype(int)

# Build vtkPolyData object, using Bane's super cool library
mesh1 = pv.PolyData(verts1, vfaces1)
mesh1['Normals'] = normals1
mesh1['values'] = values1

# Generate isosurface 2 from domain defined by bool_array2 using isovalue = -0.5
verts2, faces2, normals2, values2 = measure.marching_cubes(ellip_base, -0.5, mask=bool_array2)
vfaces2 = np.column_stack((np.ones(len(faces2),) * 3, faces2)).astype(int)

mesh2 = pv.PolyData(verts2, vfaces2)
mesh2['Normals'] = normals2
mesh2['values'] = values2

# Plot both isosurfaces
plotter = pv.Plotter()
plotter.add_mesh(mesh1, scalars='values')
plotter.add_mesh(mesh2, scalars='values')
plotter.show()
```
<img width="644" alt="Capture" src="https://user-images.githubusercontent.com/29631938/152661922-aa1b697c-aba5-4b8b-bc2b-371f4e26ad84.PNG">

**Comment by stla** [comment]
Not sure to understand. My impression is that you want `remove_points`. But `clip_scalar` is better. See https://github.com/pyvista/pyvista/issues/2026
